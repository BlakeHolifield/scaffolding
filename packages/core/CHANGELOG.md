# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

## [0.6.5](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.4...@scalprum/core-0.6.5) (2023-12-04)

## [0.6.4](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.3...@scalprum/core-0.6.4) (2023-12-04)

## [0.6.3](https://github.com/scalprum/scaffolding/compare/@scalprum/core-0.6.2...@scalprum/core-0.6.3) (2023-12-04)

# Changelog
