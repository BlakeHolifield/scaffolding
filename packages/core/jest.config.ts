/* eslint-disable */
export default {
  displayName: '@scalprum/core',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]s$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/packages/core',
};
