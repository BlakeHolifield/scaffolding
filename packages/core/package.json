{
  "name": "@scalprum/core",
  "version": "0.0.0",
  "description": "Includes core functions for scalprum scaffolding.",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "files": ["dist"],
  "typings": "dist/index.d.ts",
  "repository": "https://github.com/scalprum/scaffloding.git",
  "author": "Martin Marosi <marvusm.mmi@gmail.com>",
  "license": "Apache-2.0",
  "scripts": {
    "build": "yarn build:cjs && yarn build:esm && yarn build:typings",
    "build:esm": "tsc",
    "build:cjs": "tsc --build tsconfig.cjs.json",
    "build:typings": "node ../../scripts/create-typings.js"
  }
}
