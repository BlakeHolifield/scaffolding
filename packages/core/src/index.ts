export const GLOBAL_NAMESPACE = '__scalprum__';
export interface AppMetadata {
  name: string;
  appId?: string;
  elementId?: string;
  rootLocation?: string;
  scriptLocation?: string;
  manifestLocation?: string;
}
export interface AppsConfig {
  [key: string]: AppMetadata;
}

export type PrefetchFunction<T = any> = (ScalprumApi: Record<string, any> | undefined) => Promise<T>;
export type ExposedScalprumModule<T = any, P = any> = { default: T; prefetch?: PrefetchFunction<P> };
export type ScalprumModule<T = any, P = any> = {
  cachedModule?: ExposedScalprumModule<T, P>;
  prefetchPromise?: ReturnType<PrefetchFunction>;
};

export interface Factory<T = any, P = any> {
  init: (sharing: any) => void;
  modules: {
    [key: string]: ExposedScalprumModule<T, P>;
  };
  expiration: Date;
}

export interface ScalprumOptions {
  cacheTimeout: number;
}

export type Scalprum<T extends Record<string, any> = Record<string, any>> = {
  appsConfig: AppsConfig;
  pendingInjections: {
    [key: string]: Promise<any>;
  };
  pendingLoading: {
    [key: string]: Promise<ScalprumModule>;
  };
  pendingPrefetch: {
    [key: string]: Promise<unknown>;
  };
  factories: {
    [key: string]: Factory;
  };
  scalprumOptions: ScalprumOptions;
  api: T;
};

export type Container = Window & Factory;

declare global {
  // eslint-disable-next-line no-unused-vars
  interface Window {
    [GLOBAL_NAMESPACE]: Scalprum;
  }
}

declare function __webpack_init_sharing__(scope: string): void;
declare let __webpack_share_scopes__: any;

export const handlePrefetchPromise = (id: string, prefetch?: Promise<any>) => {
  if (prefetch) {
    setPendingPrefetch(id, prefetch);
    prefetch.finally(() => {
      removePrefetch(id);
    });
  }
};

export const getScalprum = () => window[GLOBAL_NAMESPACE];
export const getCachedModule = (scope: string, module: string, skipCache = false): ScalprumModule => {
  try {
    const factory: Factory = window[GLOBAL_NAMESPACE].factories[scope];
    if (!factory || !factory.expiration) {
      return {};
    }
    /**
     * Invalidate module after 2 minutes
     */
    const isExpired =
      skipCache || (new Date().getTime() - factory.expiration.getTime()) / 1000 > window[GLOBAL_NAMESPACE].scalprumOptions.cacheTimeout;
    if (isExpired) {
      delete window[GLOBAL_NAMESPACE].factories[scope];
      return {};
    }

    const cachedModule = factory.modules[module];
    if (!module) {
      return {};
    }

    const prefetchID = `${scope}#${module}`;
    const prefetchPromise = getPendingPrefetch(prefetchID);
    if (prefetchPromise) {
      return { cachedModule, prefetchPromise };
    }
    if (cachedModule?.prefetch) {
      handlePrefetchPromise(prefetchID, cachedModule.prefetch(getScalprum().api));
      return { cachedModule, prefetchPromise: getPendingPrefetch(prefetchID) };
    }
    return { cachedModule };
  } catch (error) {
    // If something goes wrong during the cache retrieval, reload module.
    console.warn(`Unable to retrieve cached module ${scope} ${module}. New module will be loaded.`, error);
    return {};
  }
};

export const setPendingInjection = (id: string, injectionLock: Promise<any>): void => {
  window[GLOBAL_NAMESPACE].pendingInjections[id] = injectionLock;
};

export const getPendingInjection = (id: string): Promise<any> | undefined => {
  return window[GLOBAL_NAMESPACE].pendingInjections[id];
};

export const setPendingPrefetch = (id: string, prefetch: Promise<any>): void => {
  window[GLOBAL_NAMESPACE].pendingPrefetch[id] = prefetch;
};

export const getPendingPrefetch = (id: string): Promise<any> | undefined => {
  return window[GLOBAL_NAMESPACE].pendingPrefetch?.[id];
};

export const removePrefetch = (id: string) => {
  delete window[GLOBAL_NAMESPACE].pendingPrefetch[id];
};

export const resolvePendingInjection = (id: string) => {
  delete window[GLOBAL_NAMESPACE].pendingInjections[id];
};

export const setPendingLoading = (scope: string, module: string, promise: Promise<any>): Promise<any> => {
  window[GLOBAL_NAMESPACE].pendingLoading[`${scope}#${module}`] = promise;
  promise
    .then((data) => {
      delete window[GLOBAL_NAMESPACE].pendingLoading[`${scope}#${module}`];
      return data;
    })
    .catch(() => {
      delete window[GLOBAL_NAMESPACE].pendingLoading[`${scope}#${module}`];
    });
  return promise;
};

export const getPendingLoading = (scope: string, module: string): Promise<any> | undefined => {
  return window[GLOBAL_NAMESPACE].pendingLoading[`${scope}#${module}`];
};

export const preloadModule = async (scope: string, module: string, processor?: (item: any) => string, skipCache = false) => {
  const { manifestLocation } = getAppData(scope);
  const { cachedModule } = getCachedModule(scope, module, skipCache);
  let modulePromise = getPendingLoading(scope, module);

  // lock preloading if module exists or is already being loaded
  if (!modulePromise && Object.keys(cachedModule || {}).length == 0 && manifestLocation) {
    modulePromise = processManifest(manifestLocation, scope, processor).then(() => asyncLoader(scope, module));
  }

  // add scalprum API later
  const prefetchID = `${scope}#${module}`;

  if (!getPendingPrefetch(prefetchID) && cachedModule?.prefetch) {
    handlePrefetchPromise(prefetchID, cachedModule.prefetch(getScalprum().api));
  }

  return setPendingLoading(scope, module, Promise.resolve(modulePromise));
};

export const initialize = <T extends Record<string, any> = Record<string, any>>({
  appsConfig,
  api,
  options,
}: {
  appsConfig: AppsConfig;
  api?: T;
  options?: Partial<ScalprumOptions>;
}): void => {
  const defaultOptions: ScalprumOptions = {
    cacheTimeout: 120,
    ...options,
  };
  window[GLOBAL_NAMESPACE] = {
    appsConfig,
    pendingInjections: {},
    pendingLoading: {},
    pendingPrefetch: {},
    factories: {},
    scalprumOptions: defaultOptions,
    api: api || {},
  };
};

export const getAppData = (name: string): AppMetadata => window[GLOBAL_NAMESPACE].appsConfig[name];

const shouldInjectScript = (src: string) => document.querySelectorAll(`script[src="${src}"]`)?.length === 0;

export const injectScript = (scope: string, scriptLocation: string): Promise<[any, HTMLScriptElement | undefined]> => {
  let s: HTMLScriptElement | undefined = undefined;
  if (!shouldInjectScript(scriptLocation)) {
    return Promise.resolve([scope, document.querySelectorAll(`script[src="${scriptLocation}"]`)?.[0] as HTMLScriptElement]);
  }
  const injectionPromise: Promise<[any, HTMLScriptElement | undefined]> = new Promise((res, rej) => {
    s = document.createElement('script');
    s.src = scriptLocation;
    s.id = `container_entry_${scope}`;
    s.onload = () => {
      res([scope, s]);
    };
    s.onerror = (...args) => {
      rej([args, s]);
    };
  });
  if (typeof s !== 'undefined') {
    document.body.appendChild(s);
  }

  return injectionPromise;
};

export async function processManifest(
  url: string,
  scope: string,
  processor: ((value: any) => string) | undefined
): Promise<[unknown, HTMLScriptElement | undefined][]> {
  const headers = new Headers();
  headers.append('Pragma', 'no-cache');
  headers.append('Cache-Control', 'no-cache');
  headers.append('expires', '0');
  const response = await fetch(url, {
    method: 'GET',
    headers,
  });

  // handle network errors
  if (!response.ok) {
    let error = 'Unable to process manifest';
    try {
      error = await response.json();
    } catch {
      // try text error if json fails
      error = await response.text();
    }

    throw error;
  }

  // response is OK get manifest payload
  const manifest = await response.json();
  const pendingInjection = getPendingInjection(scope);
  if (pendingInjection) {
    return pendingInjection;
  }

  const injectionPromise = Promise.all(
    Object.entries(manifest)
      .filter(([key]) => (scope ? key === scope : true))
      .flatMap(processor || ((value: any) => (value[1] as { entry: string }).entry || value))
      .map(async (scriptLocation: string) => {
        const data = await injectScript(scope, scriptLocation);
        resolvePendingInjection(scope);
        return data;
      })
  );
  setPendingInjection(scope, injectionPromise);
  injectionPromise.then((res) => {
    resolvePendingInjection(scope);
    return res;
  });
  return injectionPromise;
}

export async function asyncLoader<T = any, P = any>(scope: string, module: string): Promise<ExposedScalprumModule<T, P>> {
  if (typeof scope === 'undefined' || scope.length === 0) {
    throw new Error("Scope can't be undefined or empty");
  }
  if (typeof module === 'undefined' || module.length === 0) {
    throw new Error("Module can't be undefined or empty");
  }

  if (!module.startsWith('./')) {
    console.warn(`Your module ${module} doesn't start with './' this indicates an error`);
  }

  await __webpack_init_sharing__('default');
  const container: Container = (window as { [key: string]: any })[scope];
  await container.init(__webpack_share_scopes__.default);
  const factory = await (window as { [key: string]: any })[scope].get(module);

  if (!window[GLOBAL_NAMESPACE].factories[scope]) {
    window[GLOBAL_NAMESPACE].factories[scope] = {} as any;
  }

  const moduleFactory = factory();
  const factoryCache: Factory = {
    init: container.init,
    modules: {
      ...window[GLOBAL_NAMESPACE].factories[scope].modules,
      [module]: {
        ...moduleFactory,
      },
    },
    expiration: new Date(),
  };

  window[GLOBAL_NAMESPACE].factories[scope] = factoryCache;
  return factory();
}
