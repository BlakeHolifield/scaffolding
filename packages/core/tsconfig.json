{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist/esm",
    "declarationDir": "./dist/cjs"
  },
  "include": [
    "./src/*",
    "./src/**/*"
  ],
}
