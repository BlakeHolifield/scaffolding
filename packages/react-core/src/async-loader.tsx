import React from 'react';
import { ExposedScalprumModule, getCachedModule, getScalprum, PrefetchFunction } from '@scalprum/core';

export async function loadComponent<P = {}>(
  scope: string,
  module: string,
  importName = 'default',
): Promise<{ prefetch?: PrefetchFunction; component: React.ComponentType<P> }> {
  {
    const { pluginStore } = getScalprum();
    let mod: ExposedScalprumModule | undefined;
    const { cachedModule } = getCachedModule<React.ComponentType<P>, PrefetchFunction>(scope, module);
    mod = cachedModule;
    if (!mod) {
      mod = await pluginStore.getExposedModule<ExposedScalprumModule>(scope, module);
    }
    return {
      prefetch: mod.prefetch,
      component: mod[importName],
    };
  }
}
