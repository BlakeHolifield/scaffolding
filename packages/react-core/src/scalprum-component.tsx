import React, { useEffect, Suspense, useState, ReactNode, useReducer } from 'react';
import {
  getCachedModule,
  handlePrefetchPromise,
  getAppData,
  injectScript,
  processManifest,
  getPendingLoading,
  setPendingLoading,
  getPendingPrefetch,
} from '@scalprum/core';
import isEqual from 'lodash/isEqual';
import { loadComponent } from './async-loader';
import DefaultErrorComponent from './default-error-component';
import { PrefetchProvider } from './prefetch-context';
import { useScalprum } from './use-scalprum';

// eslint-disable-next-line @typescript-eslint/ban-types
export type ScalprumComponentProps<API extends Record<string, any> = {}, Props extends Record<string, any> = {}> = Props & {
  fallback?: NonNullable<React.ReactNode> | null;
  appName: string;
  api?: API;
  scope: string;
  module: string;
  ErrorComponent?: React.ReactElement;
  LoadingComponent?: React.ComponentType;
  innerRef?: React.Ref<unknown>;
  processor?: (item: any) => string;
  skipCache?: boolean;
};

interface LoadModuleProps extends Omit<ScalprumComponentProps, 'ErrorComponent'> {
  ErrorComponent: React.ComponentType;
}

async function setComponentFromModule(
  scope: string,
  module: string,
  isMounted: boolean,
  setComponent: React.Dispatch<
    React.SetStateAction<
      | React.ComponentType<{
          ref?: React.Ref<unknown> | undefined;
        }>
      | undefined
    >
  >
): Promise<(...args: any[]) => Promise<any> | undefined> {
  const { prefetch, component } = await loadComponent(scope, module);
  isMounted && setComponent(() => component);
  return prefetch;
}

const LoadModule: React.ComponentType<LoadModuleProps> = ({
  fallback = 'loading',
  appName,
  scope,
  module,
  ErrorComponent,
  processor,
  innerRef,
  skipCache = false,
  ...props
}) => {
  const { scriptLocation, manifestLocation } = getAppData(appName);
  const [reRender, forceRender] = useReducer((prev) => prev + 1, 0);
  const [Component, setComponent] = useState<React.ComponentType<{ ref?: React.Ref<unknown> } & Record<string, any>> | undefined>(undefined);
  const [prefetchPromise, setPrefetchPromise] = useState<Promise<any>>();
  const [loadingError, setLoadingError] = useState<Error | undefined>();

  if (loadingError) {
    // Error has to be thrown during render loop. Promise based errors won't be catched by error boundary.
    throw loadingError;
  }

  const { api: scalprumApi } = useScalprum();

  useEffect(() => {
    const prefetchID = `${scope}#${module}`;
    const { cachedModule, prefetchPromise } = getCachedModule(scope, module, skipCache);
    setPrefetchPromise(prefetchPromise);

    let isMounted = true;
    const handleLoadingError = (error?: Error) => {
      if (isMounted) {
        setLoadingError(error);
        setComponent(() => (props: any) => <ErrorComponent error={error} {...props} />);
      }
    };
    let pref;
    /**
     * Check if module is being pre-loaded
     */
    const pendingLoading = getPendingLoading(scope, module);

    if (!cachedModule && pendingLoading) {
      pendingLoading.finally(() => {
        forceRender();
      });
    } else {
      /**
       * Here will be registry check
       */
      if (!cachedModule) {
        if (scriptLocation) {
          const injectionPromise = injectScript(appName, scriptLocation)
            .then(() => {
              pref = setComponentFromModule(scope, module, isMounted, setComponent);
              if (pref) {
                pref.then((result) => {
                  if (result) {
                    const prefetch = getPendingPrefetch(prefetchID) || result(scalprumApi);
                    setPrefetchPromise(prefetch);
                    handlePrefetchPromise(prefetchID, prefetch);
                  }
                });
              }
              return pref;
            })
            .catch(handleLoadingError);

          // lock module preload
          setPendingLoading(scope, module, injectionPromise);
        } else if (manifestLocation) {
          const processPromise = processManifest(manifestLocation, appName, scope, processor)
            .then(() => {
              pref = setComponentFromModule(scope, module, isMounted, setComponent);
              pref.then((result) => {
                if (result) {
                  const prefetch = getPendingPrefetch(prefetchID) || result(scalprumApi);
                  setPrefetchPromise(prefetch);
                  handlePrefetchPromise(prefetchID, prefetch);
                }
              });
              return pref;
            })
            .catch(handleLoadingError);

          // lock module preload
          setPendingLoading(scope, module, processPromise);
        }
      } else {
        try {
          isMounted && setComponent(() => cachedModule.default);

          pref = cachedModule.prefetch;
          if (pref) {
            const prefetch = getPendingPrefetch(prefetchID) || pref(scalprumApi);
            setPrefetchPromise(prefetch);
            prefetch.then(console.error);
            handlePrefetchPromise(prefetchID, prefetch);
          }
        } catch (e) {
          handleLoadingError(e as Error);
        }
      }
    }

    return () => {
      isMounted = false;
      setLoadingError(undefined);
    };
  }, [appName, scope, skipCache, reRender]);

  // clear prefetchPromise (from factory)
  useEffect(() => {
    setPrefetchPromise(undefined);
  }, [appName, scope, module]);

  return (
    <PrefetchProvider prefetchPromise={prefetchPromise}>
      <Suspense fallback={fallback}>{Component ? <Component ref={innerRef} {...props} /> : fallback}</Suspense>
    </PrefetchProvider>
  );
};

interface BaseScalprumComponentState {
  hasError: boolean;
  error?: any;
  errorInfo?: any;
  repairAttempt?: boolean;
}

class BaseScalprumComponent extends React.Component<ScalprumComponentProps, BaseScalprumComponentState> {
  selfRepairAttempt: boolean;
  static defaultProps = {
    ErrorComponent: <DefaultErrorComponent />,
  };
  constructor(props: ScalprumComponentProps) {
    super(props);
    this.state = { hasError: false };
    this.selfRepairAttempt = false;
  }

  static getDerivedStateFromError(): BaseScalprumComponentState {
    return { hasError: true };
  }

  shouldComponentUpdate(nextProps: ScalprumComponentProps, nextState: BaseScalprumComponentState) {
    if (this.state.hasError !== nextState.hasError) {
      return true;
    }

    return !isEqual(nextProps, this.props) || !isEqual(nextState, this.state);
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    if (this.selfRepairAttempt === true) {
      console.error('Scalprum encountered an error!', error.message);
      this.setState({ error, errorInfo });
    } else {
      console.warn('Scalprum failed to render component. Attempting to skip module cache.');
      this.setState({ repairAttempt: true });
    }
  }

  render(): ReactNode {
    const { ErrorComponent = <DefaultErrorComponent {...this.state} />, ...props } = this.props;

    const PreparedError: React.ComponentType = (props: any) => {
      return React.cloneElement<typeof this.state>(ErrorComponent, { ...this.state, ...props });
    };

    if (this.state.repairAttempt && !this.selfRepairAttempt) {
      /**
       * Retry fetching module with disabled cache
       */
      this.selfRepairAttempt = true;
      return <LoadModule {...props} skipCache ErrorComponent={PreparedError} />;
    }

    if (this.state.hasError && this.selfRepairAttempt) {
      return React.cloneElement(ErrorComponent as React.FunctionComponentElement<BaseScalprumComponentState>, { ...this.state });
    }
    return <LoadModule {...props} ErrorComponent={PreparedError} />;
  }
}

export const ScalprumComponent: React.ComponentType<ScalprumComponentProps> = React.forwardRef((props, ref) => (
  <BaseScalprumComponent {...props} innerRef={ref} />
));
