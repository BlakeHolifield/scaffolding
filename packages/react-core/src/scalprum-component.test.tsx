import React from 'react';
import * as asyncComponent from './async-loader';
import { ScalprumComponent } from './scalprum-component';
import { render, cleanup, act } from '@testing-library/react';
import * as ScalprumCore from '@scalprum/core';
import { AppsConfig, GLOBAL_NAMESPACE } from '@scalprum/core';

describe('<ScalprumComponent />', () => {
  const mockInitScalprumConfig: AppsConfig = {
    appOne: {
      name: 'appOne',
      appId: 'appOne',
      rootLocation: '/foo',
      scriptLocation: '/bar.js',
      elementId: 'id',
    },
  };

  const mockInitScalpumConfigManifest: AppsConfig = {
    appOne: {
      name: 'appOne',
      appId: 'appOne',
      rootLocation: '/foo',
      manifestLocation: '/bar.json',
      elementId: 'id',
    },
  };
  const getAppDataSpy = jest.spyOn(ScalprumCore, 'getAppData').mockReturnValue(mockInitScalprumConfig.appOne);
  const injectScriptSpy = jest.spyOn(ScalprumCore, 'injectScript');
  const processManifestSpy = jest.spyOn(ScalprumCore, 'processManifest');
  const loadComponentSpy = jest.spyOn(asyncComponent, 'loadComponent').mockReturnValue(() => import('./TestComponent'));

  afterEach(() => {
    cleanup();
    window[GLOBAL_NAMESPACE] = undefined;
    getAppDataSpy.mockClear();
    injectScriptSpy.mockClear();
    processManifestSpy.mockClear();
  });

  test('should retrieve script location', () => {
    ScalprumCore.initialize({ scalpLets: mockInitScalprumConfig });
    ScalprumCore.setPendingInjection('appOne', jest.fn());
    ScalprumCore.initializeApp({ name: 'appOne', id: 'id', mount: jest.fn(), unmount: jest.fn(), update: jest.fn() });
    render(<ScalprumComponent appName="appOne" scope="some" module="test" />);

    expect(getAppDataSpy).toHaveBeenCalledWith('appOne');
  });

  test('should retrieve manifest location', () => {
    getAppDataSpy.mockReturnValueOnce(mockInitScalpumConfigManifest.appOne);
    ScalprumCore.initialize({ scalpLets: mockInitScalpumConfigManifest });
    ScalprumCore.setPendingInjection('appOne', jest.fn());
    ScalprumCore.initializeApp({ name: 'appOne', id: 'id', mount: jest.fn(), unmount: jest.fn(), update: jest.fn() });
    render(<ScalprumComponent appName="appOne" scope="some" module="test" />);

    expect(getAppDataSpy).toHaveBeenCalledWith('appOne');
  });

  test('should inject script and mount app if it was not initialized before', async () => {
    const mount = jest.fn();
    ScalprumCore.initialize({ scalpLets: mockInitScalprumConfig });
    injectScriptSpy.mockImplementationOnce(() => {
      ScalprumCore.setPendingInjection('appOne', jest.fn());
      ScalprumCore.initializeApp({ name: 'appOne', mount, unmount: jest.fn(), update: jest.fn(), id: 'appOne' });
      return Promise.resolve(['', undefined]);
    });
    await act(async () => {
      render(<ScalprumComponent appName="appOne" scope="some" module="test" />);
    });

    expect(mount).toHaveBeenCalled();
    expect(injectScriptSpy).toHaveBeenCalledWith('appOne', '/bar.js');
  });

  test('should inject manifest and mount app if it was not initialized before', async () => {
    getAppDataSpy.mockReturnValueOnce(mockInitScalpumConfigManifest.appOne);
    const mount = jest.fn();
    ScalprumCore.initialize({ scalpLets: mockInitScalpumConfigManifest });
    processManifestSpy.mockImplementationOnce(() => {
      ScalprumCore.setPendingInjection('appOne', jest.fn());
      ScalprumCore.initializeApp({ name: 'appOne', mount, unmount: jest.fn(), update: jest.fn(), id: 'appOne' });
      return Promise.resolve([['', undefined]]);
    });
    await act(async () => {
      render(<ScalprumComponent appName="appOne" scope="some" module="test" />);
    });

    expect(mount).toHaveBeenCalled();
    expect(processManifestSpy).toHaveBeenCalledWith('/bar.json', 'appOne', 'some', undefined);
  });

  test('should not inject script the app was initialized before', async () => {
    const mount = jest.fn();
    ScalprumCore.initialize({ scalpLets: mockInitScalprumConfig });
    ScalprumCore.setPendingInjection('appOne', jest.fn());
    ScalprumCore.initializeApp({ name: 'appOne', mount, unmount: jest.fn(), update: jest.fn(), id: 'appOne' });
    await act(async () => {
      render(<ScalprumComponent appName="appOne" scope="some" module="test" />);
    });

    expect(mount).toHaveBeenCalled();
    expect(injectScriptSpy).not.toHaveBeenCalled();
    expect(processManifestSpy).not.toHaveBeenCalled();
  });

  test('should not process manifest the app was initialized before', async () => {
    getAppDataSpy.mockReturnValueOnce(mockInitScalpumConfigManifest.appOne);
    const mount = jest.fn();
    ScalprumCore.initialize({ scalpLets: mockInitScalpumConfigManifest });
    ScalprumCore.setPendingInjection('appOne', jest.fn());
    ScalprumCore.initializeApp({ name: 'appOne', mount, unmount: jest.fn(), update: jest.fn(), id: 'appOne' });
    await act(async () => {
      render(<ScalprumComponent appName="appOne" scope="some" module="test" />);
    });

    expect(mount).toHaveBeenCalled();
    expect(processManifestSpy).not.toHaveBeenCalled();
    expect(injectScriptSpy).not.toHaveBeenCalled();
  });

  test('should render test component', async () => {
    const mount = jest.fn();
    ScalprumCore.initialize({ scalpLets: mockInitScalprumConfig });
    injectScriptSpy.mockImplementationOnce(() => {
      ScalprumCore.setPendingInjection('appOne', jest.fn());
      ScalprumCore.initializeApp({ name: 'appOne', mount, unmount: jest.fn(), update: jest.fn(), id: 'appOne' });
      return Promise.resolve(['', undefined]);
    });
    let container;
    await act(async () => {
      container = render(<ScalprumComponent appName="appOne" scope="some" module="test" />)?.container;
    });

    expect(loadComponentSpy).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  test('should render test component with manifest', async () => {
    getAppDataSpy.mockReturnValueOnce(mockInitScalpumConfigManifest.appOne);
    const mount = jest.fn();
    ScalprumCore.initialize({ scalpLets: mockInitScalprumConfig });
    processManifestSpy.mockImplementationOnce(() => {
      ScalprumCore.setPendingInjection('appOne', jest.fn());
      ScalprumCore.initializeApp({ name: 'appOne', mount, unmount: jest.fn(), update: jest.fn(), id: 'appOne' });
      return Promise.resolve([['', undefined]]);
    });
    let container;
    await act(async () => {
      container = render(<ScalprumComponent appName="appOne" scope="some" module="test" />)?.container;
    });

    expect(loadComponentSpy).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
