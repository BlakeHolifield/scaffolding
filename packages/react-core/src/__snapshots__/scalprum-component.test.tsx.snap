// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScalprumComponent /> should render test component 1`] = `
<div>
  <span>
    Test
  </span>
</div>
`;

exports[`<ScalprumComponent /> should render test component with manifest 1`] = `
<div>
  <span>
    Test
  </span>
</div>
`;
